AirCasting/ABConnector/BluetoothManager.swift

```diff
@@ -68,6 +75,49 @@ class BluetoothManager: NSObject, ObservableObject {
             let _ = centralManager
         }
     }
+
+    // MARK: - Refactored part
+    // This is the part of this class that is already refactored.
+
+    typealias CharacteristicObserverAction = (Result<Data?, Error>) -> Void
+
+    private struct CharacteristicObserver {
+        let identifier = UUID()
+        let action: CharacteristicObserverAction
+    }
+
+    // The mapping is CBUUID -> CharacteristicObserver
+    // and not UUID(observation token) -> CharacteristicObserver
+    // because we care more about the update performance than un-register performance.
+    private var charactieristicsMapping: [CBUUID: [CharacteristicObserver]] = [:]
+    private let characteristicsMappingLock = NSLock()
+
+    /// Adds an entry to observers of a particular characteristic
+    /// - Parameters:
+    ///   - characteristic: UUID of characteristic to observe
+    ///   - notify: block called each time characteristic changes (either changes value or throws an error)
+    /// - Returns: Opaque token to use when un-registering
+    func subscribeToCharacteristic(_ characteristic: CBUUID, notify: @escaping CharacteristicObserverAction) -> AnyHashable {
+        let observer = CharacteristicObserver(action: notify)
+        characteristicsMappingLock.lock()
+        charactieristicsMapping[characteristic, default:[]].append(observer)
+        characteristicsMappingLock.unlock()
+        return observer.identifier
+    }
+
+    /// Removes an entry from observing characteristic
+    /// - Parameter token: Opaque token received on subscription
+    /// - Returns: A `Bool` value indicating if a given token was successfuly removed. Only reason it can fail is double unregistration.
+    @discardableResult func unsubscribeCharacteristicObserver(_ token: AnyHashable) -> Bool {
+        guard let uuid = token as? UUID else { return false }
+//        characteristicsMappingLock.lock()
```

Comment:
I think we should switch from `NSLock` to `NSRecursiveLock` and should be fine. Can you please test it out? Also there is a missed return without unlocking right in the below, so that's not good. Let's do:
`characteristicsMappingLock.lock(); defer { characteristicsMappingLock.unlock() }` (and remove the commented unlock from L118)